### PR TITLE
[dontmerge]: Change `setenv` on Linux to panic if threaded

### PR DESCRIPTION
**Pull request by cgwalters**
See this thread
https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/

My main goal here is to have this go through all of Rust's CI
and see if any panics turn up.  There are a variety of `set_env()`
calls, but AFAICS they are all early in main.

Then if that works, I'd like to do a crater run so we can gauge
the fallout in the ecosystem of how many places need to be fixed
for this.

(I think we should introduce `try_set_env()`, but this will still
 help us understand the fallout)

### CODE REVIEW

**Comment by cgwalters** [comment]
Explicitly marking draft since obviously this should not merge.  But let's get CI to run for it.

**Comment by cgwalters** [comment]
> tidy error: /checkout/library/std/src/sys/unix/os.rs:547: TODO is deprecated; use FIXME

Hah, not the CI failure I was expecting...just dropped that bit.

File: library/std/src/sys/unix/os.rs [hunk 1]
```diff
@@ -543,7 +543,21 @@ pub fn getenv(k: &OsStr) -> Option<OsString> {
     }
 }
 
+fn process_is_threaded() -> bool {
+    if cfg!(any(target_os = "linux", target_os = "android")) {
+        if let Ok(r) = crate::fs::read_dir("/proc/self/task") {
```
Comment (bjorn3) [inline library/std/src/sys/unix/os.rs:548]: Rsix checks that `/proc` is an actual proc fs: https://github.com/bytecodealliance/rustix/blob/3e20a74673b1faa3d2c9bf7f4182a68173ef3532/src/io/procfs.rs#L199-L206

**Comment by cgwalters** [comment]
> ---- process::tests::test_capture_env_at_spawn stdout ----
> thread 'process::tests::test_capture_env_at_spawn' panicked at 'Tried to setenv > '"RUN_TEST_NEW_ENV2"' while process has multiple threads', library/std/src/sys/unix/os.rs:558:9

Ah neat, this turned up a test case that fails because cargo defaults to running tests in threads.  Now the interesting question is whether these tests are linking in and invoking any C code which might ultimately invoke `getenv()`.  *Probably* not...but knowing for sure is not trivial.

And now that I look there's of course also `library/std/tests/env.rs`.  

I guess the question of what to do with these tests is a microcosm of the larger issue.  We could add something like `#[test(process)]` that spawned them as a (single threaded) subprocess perhaps?

**Comment by bjorn3** [comment]
libtest supports spawning subprocesses for each thread when using `-Zpanic-abort-tests`, but I believe the current implementation can sometimes deadlock.

**Comment by jyn514** [comment]
> I guess the question of what to do with these tests is a microcosm of the larger issue. We could add something like #[test(process)] that spawned them as a (single threaded) subprocess perhaps?

I maintain https://github.com/deadlinks/cargo-deadlinks which uses setenv in a non-trivial number of tests. I personally don't mind updating them, but if you multiply across the whole ecosystem I doubt just making this an unconditional panic is feasible.

**Comment by jyn514** [comment]
Note also that using a separate process for each test adds overhead, so I don't think changing that to be the default is a great solution either (and even so, it wouldn't fix set_env outside of test code).

File: library/std/src/sys/unix/os.rs (hunk 1 above)
Comment (cgwalters) [reply library/std/src/sys/unix/os.rs:548]: OK yeah...the rustix code is really nice I have to say.

But, we'll cross the "serious implementation of `process_is_threaded()`" bridge when(if?) there is agreement to do a PR adding `try_set_env()` to the std.

**Comment by RalfJung** [comment]
> Ah neat, this turned up a test case that fails because cargo defaults to running tests in threads.

That makes all tests that mutate the environment rather dubious, doesn't it? They could interfere with other threads that also access the environment. Even if everything is synchronized, this can quickly lead to logic errors. Basically this requires each test to have its own private environment variable that no other test may use.

**Comment by cgwalters** [comment]
> That makes all tests that mutate the environment rather dubious, doesn't it? 

Now that I think about this a bit more, note the failing tests here are basically tests for `set_env()` itself.  So I would say this isn't actually reflective of the wider ecosystem or problem domain.

**Comment by jyn514** [comment]
@craterbot check

**Comment by jyn514** [comment]
@craterbot check start=master#62efba8a050c64249dab942951bb28f710208bc8 end=try#8262275854dc0e5381c9d42d56257d94932c5a87

**Comment by the8472** [comment]
Shouldn't that be a build-test crater run? Check won't catch panics.

**Comment by jyn514** [comment]
@craterbot cancel

**Comment by jyn514** [comment]
@craterbot abort

**Comment by jyn514** [comment]
@craterbot run mode=build-and-test  start=master#62efba8a050c64249dab942951bb28f710208bc8 end=try#8262275854dc0e5381c9d42d56257d94932c5a87

**Comment by Mark-Simulacrum** [comment]
@craterbot retry-report

**Comment by cgwalters** [comment]
Sorry, I got distracted by other things.  I don't have a strong opinion, we can probably close this because I got the feedback I wanted.  (Actually, just doing that)

But we should probably take more discussion to https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/

**Pull request closed without merging**